File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] fq = new int[26];
        int mostFreqLetter = 0;
        int left = 0;
        int right = 0;
        int res = 0;

        while(right < s.length()) {
            // add char count to array
            fq[s.charAt(right) - 'A']++;

            // update the most frequent letter count 
            mostFreqLetter = Math.max(mostFreqLetter, fq[s.charAt(right) - 'A']);

            int windowSize = right - left + 1;

            if((windowSize - mostFreqLetter) > k) {
                fq[s.charAt(left) - 'A']--;
                left++;
            }

            right++;
        }

        res = Math.min(s.length(), mostFreqLetter + k);
        return res;
    }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        counts = {}
        res = 0

        l = 0
        for r in range(len(s)):
            counts[s[r]] = 1 + counts.get(s[r], 0)
            while (r - l + 1) - max(counts.values()) > k:
                counts[s[l]] -= 1
                l += 1
            res = max(res, (r-l+1))

        return res



